infer/src/pulse/PulseFormula.mli

```diff
@@ -18,11 +18,97 @@ module Var :
      and module Set = PulseAbstractValue.Set
      and module Map = PulseAbstractValue.Map
 
+(* Pulse-infinite added *)
+
+(** Linear Arithmetic *)
+module LinArith : sig
+  (** linear combination of variables, eg [2·x + 3/4·y + 12] *)
+  type t [@@deriving compare, yojson_of, equal]
+end
+
+module Term : sig
+  type t =
+    | Const of Q.t
+    | String of string
+    | Var of Var.t
+    | Procname of Procname.t
+    | FunctionApplication of {f: t; actuals: t list}
+    | Linear of LinArith.t
+    | Add of t * t
+    | Minus of t
+    | LessThan of t * t
+    | LessEqual of t * t
+    | Equal of t * t
+    | NotEqual of t * t
+    | Mult of t * t
+    | DivI of t * t
+    | DivF of t * t
+    | And of t * t
+    | Or of t * t
+    | Not of t
+    | Mod of t * t
+    | BitAnd of t * t
+    | BitOr of t * t
+    | BitNot of t
+    | BitShiftLeft of t * t
+    | BitShiftRight of t * t
+    | BitXor of t * t
+    | StringConcat of t * t
+    | IsInstanceOf of {var: Var.t; typ: Typ.t; nullable: bool}
+    | IsInt of t
+  [@@deriving compare, equal, yojson_of]
+
+  module Set : Stdlib.Set.S [@@deriving compare]
+end
+
+module Atom : sig
+  type t =
+    | LessEqual of Term.t * Term.t
+    | LessThan of Term.t * Term.t
+    | Equal of Term.t * Term.t
+    | NotEqual of Term.t * Term.t
+  [@@deriving compare, equal, yojson_of]
+
+  val equal : Term.t -> Term.t -> t
+
+  module Set : Stdlib.Set.S [@@deriving compare]
+
+  module Map : Stdlib.Map.S [@@deriving compare]
+end
+
+module Formula : sig
+  type t [@@deriving compare, equal, yojson_of]
+end
+
+type t =
+  { conditions: int Atom.Map.t
+        (** collection of conditions that have been assumed (via [PRUNE] CFG nodes) along the path.
+            Note that these conditions are *not* normalized w.r.t. [phi]: [phi] already contains
+            them so normalization w.r.t. [phi] would make them trivially true most of the time. *)
+  ; phi: Formula.t
+        (** the arithmetic constraints of the current symbolic state; true in both the pre and post
+            since abstract values [Var.t] have immutable semantics *) }
+[@@deriving compare, equal, yojson_of]
+
+val extract_path_cond : t -> int Atom.Map.t
+
+val extract_terminal_atoms : t -> Atom.Set.t
+
+val extract_terminal_terms : t -> Term.Set.t
+
+val map_is_empty : int Atom.Map.t -> bool
+
+val set_is_empty : Atom.Set.t -> bool
+
+val termset_is_empty : Term.Set.t -> bool
```

Comment:
I think we can avoid revealing the internal structure of formulas by just defining type aliases to what is needed here and keeping these opaque in the .mli. So in the .ml have `type 'a atom_map = 'a Atom.Map.t`, `type atom_set = Atom.Set.t`, and `type term_set = Term.Set.t`, and here in the mli just have the following (and delete everything else added above).
```suggestion
type _ atom_map
type atom_set
type term_set

val extract_path_cond : t -> int atom_map

val extract_terminal_atoms : t -> atom_set

val extract_terminal_terms : t -> term_set

val map_is_empty : _ atom_map -> bool

val set_is_empty : atom_set -> bool

val termset_is_empty : atom_set -> bool
```